src/zkevm_specs/evm/execution/end_block.py

```diff
@@ -4,69 +4,98 @@
 from typing import Set
 
 # EndBlock is an execution state that constraints the following:
-# 1. Once the EndBlock state is reached, there's no other execution states appearing until the end of the EVM Circuit.  In particular, after the first EndBlock, there will be no new lookups to the rw_table.
-# 2. The number of meaningful entries (non-padding) in the rw_table match the rw_counter after the EndBlock state is processed.
-# 3. The number of txs processed by the EVM Circuit match the number of txs in the TxTable
-#    total_tx = instruction.call_context_lookup(CallContextFieldTag.TxId)
-
+# A. Once the EndBlock state is reached, there's no other execution states appearing until the end of the EVM Circuit.  In particular, after the first EndBlock, there will be no new lookups to the rw_table.
+# B. The number of meaningful entries (non-padding) in the rw_table match the rw_counter after the EndBlock state is processed.
+# C. The number of txs processed by the EVM Circuit match the number of txs in the TxTable
+#
 # As an extra point:
-# 4. We need to prove that the EndBlock state exists
-
-# We prove (1) by constraining the transition rule that after an EndBlock
+# D. We need to prove that at least one EndBlock state exists
+#
+# We prove (A) by constraining the transition rule that after an EndBlock
 # state, only an EndBlock state can follow.
 #
-# We prove (2) and (3) by proving that at least `MAX_COUNT - evm_circuit_count`
-# padding elements exist in the rw table and tx table.  To achieve this, we do
-# padding at the beginning of both tables with sequential indexes at
-# `rw_counter` (in rw_table) or `tx_id` (in tx_table).
+# We prove (B) by proving that at least `MAX_RWS - total_rws` padding elements
+# exist in the rw table.  To achieve this, we do padding at the beginning of
+# the RwTable with a sequential index `rw_counter`.  The RwTable constraints
+# that padding rows have sequential indexes, so by doing 2 lookups with
+# rw_counter = {a, b} we know there are at least b-a padding rows.
+#
+# We prove (C) by showing that the transaction ID after the last one processed
+# corresponds to a padding tx in the TxTable, which enforces that once a
+# padding txs appears in the table, the rest will also be padding.  This way,
+# we know that if a padding tx with tx_id exist, then at most there are tx_id-1
+# non-padding txs.  In case we have exhausted the TxTable with txs, we won't
+# have any padding txs; so we skip the lookup.
 #
-# We prove (4) in the circuit implementation by constraining that the last
+# We prove (D) in the circuit implementation by constraining that the last
```

Comment:
Yes!  The reason is that if we have an empty block, there will be no transactions processed in the EVM circuit, and we need to prove that the only situation in which this can occur is when the TxTable is full of padding txs.  And we do that in the EndBlock.  Otherwise we may have a TxTable with k transactions, the EVM circuit empty (with no step), and all constraints would pass!